compiler/rustc_hir_typeck/src/expr.rs

```diff
@@ -2349,24 +2350,39 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
-                    if let Some((index, field)) = fields
-                        .iter_enumerated()
-                        .find(|(_, f)| f.ident(self.tcx).normalize_to_macros_2_0() == ident)
-                    {
+                    let mut adt_def = *base_def;
+                    let mut last_ty = None;
+                    let mut nested_fields = Vec::new();
+                    let mut index = None;
+                    while let Some(idx) = self.tcx.find_field((adt_def.did(), ident)) {
+                        let &mut first_idx = index.get_or_insert(idx);
+                        let field = &adt_def.non_enum_variant().fields[idx];
                         let field_ty = self.field_ty(expr.span, field, args);
-                        // Save the index of all fields regardless of their visibility in case
-                        // of error recovery.
-                        self.write_field_index(expr.hir_id, index);
-                        let adjustments = self.adjust_steps(&autoderef);
-                        if field.vis.is_accessible_from(def_scope, self.tcx) {
-                            self.apply_adjustments(base, adjustments);
-                            self.register_predicates(autoderef.into_obligations());
+                        if let Some(ty) = last_ty {
+                            nested_fields.push((ty, idx));
+                        }
+                        if field.ident(self.tcx).normalize_to_macros_2_0() == ident {
+                            // Save the index of all fields regardless of their visibility in case
+                            // of error recovery.
+                            self.write_field_index(expr.hir_id, first_idx, nested_fields);
+                            let adjustments = self.adjust_steps(&autoderef);
+                            if field.vis.is_accessible_from(def_scope, self.tcx) {
+                                self.apply_adjustments(base, adjustments);
+                                self.register_predicates(autoderef.into_obligations());
 
-                            self.tcx.check_stability(field.did, Some(expr.hir_id), expr.span, None);
-                            return field_ty;
+                                self.tcx.check_stability(
+                                    field.did,
+                                    Some(expr.hir_id),
+                                    expr.span,
+                                    None,
+                                );
+                                return field_ty;
+                            }
+                            private_candidate = Some((adjustments, base_def.did()));
+                            break;
                         }
-                        private_candidate = Some((adjustments, base_def.did()));
+                        last_ty = Some(field_ty);
+                        adt_def = field_ty.ty_adt_def().expect("expect Adt for unnamed field");
```

Comment:
I'm a little confused by this.

As best I can understand, for some field `a` in an anonymous ADT, you're expecting `find_field` to return the anonymous ADT's index, and then when that doesn't match the `ident` you asked for, you go into the next iteration of the loop with the anonymous ADT, where you will find `a`. In doing this, you accumulate into `nested_fields`, which gets written into the tables.

In an example like this..

```rust
struct Foo {
    a: u32,
    _: union { // Anon1
        b: u32,
        c: u32,
        _: struct { // Anon2
            d: u32,
            e: u32,
        },
    },
}
```

..then you end up with `first_idx=1` and `nested_fields=[(Anon1, 3), (Anon2, 2)]`, for `foo.e`, right?

Reading `find_field`, isn't it going to return the index of `a` in the anonymous ADT? Since it recurses? I might be missing something but doesn't this just always hit the `break` on the first iteration? It must not because your `mir-opt` tests seem correct.. but I can't see what I'm missing.

Reply:
I agree that `find_field` is a little bit confusing, and I've also posted another possibility of implementing this query [here](https://github.com/rust-lang/rust/pull/115367#issuecomment-1879633740). That is, to use this query instead

```
query nested_field_map(def_id: DefId) -> &'tcx FxHashMap<rustc_span::symbol::Ident, rustc_target::abi::FieldIdx>;
```
This query returns a map from all its field idents (including nested fields) to its relative index of the outermost ADT. In your example,
- query on `Foo` should return `hash_map!("a" => 0, "b" => 1, "c" => 1, "d" => 1, "e" => 1)`;
- query on `Foo::Anon1` should return `hash_map!("b" => 0, "c" => 1, "d"  => 2, "e" => 2)`; 
- query on `Foo::Anon1::Anon2` should return `hash_map!("d" => 0, "e" => 1)`.

**Pros**:
- going through each ADT only once, we can collect all the indexing information of every nested field in it.

**Cons**:
- the `FxHashMap` keeps more memory in `TyCtxt`;
- it is wasty if we build a `FxHashMap` for every ADTs even if they don't contain unnamed fields (though this can be avoided by opting them out).